crypto/cms/cms_sd.c

```diff
@@ -227,6 +227,30 @@ int ossl_cms_SignerIdentifier_cert_cmp(CMS_SignerIdentifier *sid, X509 *cert)
         return -1;
 }
 
+/* Method to map any, incl. provider-implemented PKEY types to OIDs */
+static int cms_generic_sign(CMS_SignerInfo *si, int verify)
+{
+    if (!verify) {
+        int hnid;
+        X509_ALGOR *alg1, *alg2;
+        EVP_PKEY *pkey = si->pkey;
+        const char *typename = EVP_PKEY_get0_type_name(pkey);
+
+        CMS_SignerInfo_get0_algs(si, NULL, NULL, &alg1, &alg2);
+        if (alg1 == NULL || alg1->algorithm == NULL)
+            return -1;
+        hnid = OBJ_obj2nid(alg1->algorithm);
+        if (hnid == NID_undef)
+            return -1;
+
+        /* Generic PKEY->name->NID->OID mapping check */
+        if (typename == NULL)
+            return -1;
+        return X509_ALGOR_set0(alg2, OBJ_txt2obj(typename, 0), V_ASN1_UNDEF, NULL);
```

Comment:
I might have instead expected there to be a call to `OBJ_txt2nid(typename)` and then followed by a call to `OBJ_find_sigid_by_algs` to find the oid for the combined sig alg/hash.

In fact this generic flow might even work for ECDSA/DSA too.

Reply:
> In fact this generic flow might even work for ECDSA/DSA too.

I thought so too (the generic code is "modeled" after the ECDSA code) but wanted to keep the change to a minimum.

Reply:
Understandable. I think it boils down to me understanding how to tweak the STACK_OF `find` logic to match/accept empty (registered) digests (NID_undef) with some (provided, non-undef) digest NID... Some code-reading and understanding required. If done, the cms_ecdsa_sign code can go.

Reply:
So done in https://github.com/openssl/openssl/pull/17733/commits/35cf0e79cf13ee24909dc281d177562eb7dc93e2